packages/flutter/lib/src/widgets/selectable_region.dart

```diff
@@ -2834,3 +2882,282 @@ typedef SelectableRegionContextMenuBuilder = Widget Function(
+
+/// A [SelectionContainer] that allows the user to listen to selection changes
+/// for the child subtree it wraps under a [SelectionArea] or [SelectableRegion].
+///
+/// This widget should have an ancestor [SelectionArea] or [SelectableRegion]
+/// to be able to listen to selection changes in this widgets subtree.
+///
+/// This widget does not listen to selection changes of nested [SelectionArea]s
+/// or [SelectableRegion]s in its subtree because those widgets create their own
+/// [SelectionRegistrar]s that do not register to any ancestor registrar.
+///
+/// {@tool dartpad}
+/// This example shows how to color red the active selection
+/// under a [SelectionArea] or [SelectableRegion].
+///
+/// ** See code in examples/api/lib/material/selection_area/selection_area.1.dart **
+/// {@end-tool}
+///
+/// {@tool dartpad}
+/// This example shows how to replace the active selection
+/// under a [SelectionArea] or [SelectableRegion] with a widget.
+///
+/// ** See code in examples/api/lib/material/selection_area/selection_area.2.dart **
+/// {@end-tool}
+///
+/// See also:
+///
+///   * [SelectionArea], which provides an overview of the selection system.
+///   * [SelectableRegion], which provides an overview of the selection system.
+class SelectionListener extends StatefulWidget {
+  /// Create a new [SelectionListener] widget.
+  const SelectionListener({
+    super.key,
+    required this.onSelectionChanged,
+    required this.child,
+  });
+
+  /// Called when the user changes the selection of children selectables
+  /// registered to its local [SelectionRegistrar].
+  final SelectionListenerSelectionChangedCallback onSelectionChanged;
+
+  /// The child widget this selection listener applies to.
+  ///
+  /// {@macro flutter.widgets.ProxyWidget.child}
+  final Widget child;
+
+  @override
+  State<SelectionListener> createState() => _SelectionListenerState();
+}
+
+class _SelectionListenerState extends State<SelectionListener> {
+  late final _SelectionListenerDelegate _selectionDelegate;
+
+  @override
+  void initState() {
+    super.initState();
+    _selectionDelegate = _SelectionListenerDelegate(onSelectionChanged: widget.onSelectionChanged);
+  }
+
+  @override
+  void dispose() {
+    _selectionDelegate.dispose();
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return SelectionContainer(
+      delegate: _selectionDelegate,
+      child: widget.child,
+    );
+  }
+}
+
+class _SelectionListenerDelegate extends MultiSelectableSelectionContainerDelegate {
+  _SelectionListenerDelegate({required this.onSelectionChanged});
+
+  final SelectionListenerSelectionChangedCallback onSelectionChanged;
+
+  @override
+  SelectionResult dispatchSelectionEvent(SelectionEvent event) {
+    final SelectionResult result = super.dispatchSelectionEvent(event);
+    onSelectionChanged(getSelections());
```

Comment:
Mainly just copied code from `_SelectableRegionContainerDelegate`. I can extend it since `SelectionListener` is in the same file, but if we decide to have it in its own file this code will need to be copied for now. I have a PR out to make a public version of `_SelectableRegionContainerDelegate`, but haven't had a chance to get back to it https://github.com/flutter/flutter/pull/147080. When that is landed I can have this container delegate extend from that as well.